tests/unit/dfem/test_mass.cpp

```diff
@@ -51,37 +51,114 @@ void DFemMass(const char *filename, int p, const int r)
+                              tuple{ Value<U>{} },
+                              *ir, all_domain_attr);
+      dop.SetParameters({ nodes });
+
+      fes.GetRestrictionMatrix()->Mult(x, X);
+      dop.Mult(X, Z);
+      Y -= Z;
+
+      real_t norm_g, norm_l = Y.Normlinf();
+      MPI_Allreduce(&norm_l, &norm_g, 1, MPI_DOUBLE, MPI_MAX, pmesh.GetComm());
+      REQUIRE(norm_g == MFEM_Approx(0.0));
+      MPI_Barrier(MPI_COMM_WORLD);
+   }
+
+   // Test boundary
+   // This ensures that we're not trying to test on fully periodic meshes
+   if (!((std::string("../../data/periodic-square.mesh").compare(filename) == 0) ||
+         (std::string("../../data/periodic-cube.mesh").compare(filename) == 0)))
+   {
+      SECTION("boundary")
+      {
+         const auto *ir = &IntRules.Get(pmesh.GetTypicalFaceGeometry(), 2 * p + r);
+
+         Array<int> all_bdr_attr;
+         if (pmesh.bdr_attributes.Size() > 0)
+         {
+            all_bdr_attr.SetSize(pmesh.bdr_attributes.Max());
+            all_bdr_attr = 1;
+         }
+
+         ParBilinearForm blf(&fes);
+         blf.AddBoundaryIntegrator(new MassIntegrator(one, ir));
+         blf.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+         blf.Assemble();
+         blf.Mult(x, y);
+         fes.GetProlongationMatrix()->MultTranspose(y, Y);
+
+         static constexpr int U = 0, Coords = 1;
+         const auto sol = std::vector{FieldDescriptor{U, &fes}};
+         DifferentiableOperator dop(sol, {{Coords, nodes->ParFESpace()}}, pmesh);
+
+         const auto mf_mass_qf =
+            [] MFEM_HOST_DEVICE(const dscalar_t &u,
+                                const tensor<real_t, DIM, BDIM> &J,
+                                const real_t &w)
+         {
+            return tuple{u * weight(J) * w};
+         };
+
+         auto derivatives = std::integer_sequence<size_t, U> {};
+         dop.AddBoundaryIntegrator(mf_mass_qf,
+                                   tuple{ Value<U>{}, Gradient<Coords>{}, Weight{} },
+                                   tuple{ Value<U>{} },
+                                   *ir, all_bdr_attr, derivatives);
+         dop.SetParameters({nodes});
+
+         fes.GetRestrictionMatrix()->Mult(x, X);
+         dop.Mult(X, Z);
+
+         Y -= Z;
+         real_t norm_g, norm_l = Y.Normlinf();
+         MPI_Allreduce(&norm_l, &norm_g, 1, MPI_DOUBLE, MPI_MAX, pmesh.GetComm());
+         REQUIRE(norm_g == MFEM_Approx(0.0));
+
+         auto dRdU = dop.GetDerivative(U, {&x}, {nodes});
+         dRdU->Mult(X, Z);
+
+         fes.GetProlongationMatrix()->MultTranspose(y, Y);
+         Y -= Z;
+         norm_l = Y.Normlinf();
+         MPI_Allreduce(&norm_l, &norm_g, 1, MPI_DOUBLE, MPI_MAX, pmesh.GetComm());
+         REQUIRE(norm_g == MFEM_Approx(0.0));
+         MPI_Barrier(MPI_COMM_WORLD);
+      }
+   }
 }
 
-TEST_CASE("DFEM Mass", "[Parallel][DFEM]")
+TEST_CASE("DFEM Mass", "[Parallel][dFEM Mass]")
```

Comment:
```suggestion
TEST_CASE("dFEM Mass", "[Parallel][dFEM]")
```
We should synchronize with the updated tags from #5022